File: src/robot_life/behavior/decay_tracker.py
# Behavior Decay & Randomness Tracker

from __future__ import annotations

import random
from collections import defaultdict
from time import monotonic


class BehaviorDecayTracker:
    """Track repeated interactions to implement response decay and randomness.

    PRD §5.1: "同人同场景反应强度逐渐降低", "30%+ 场景仅做非语音反馈".

    For the same (scene_type, target_id) pair, the tracker:
    1. Counts how many times we've responded within a decay window.
    2. Returns a ``strength`` multiplier (1.0 → 0.3) that decays with repetition.
    3. Recommends whether to use voice or silent (visual-only) feedback.
    """

    def __init__(
        self,
        *,
        decay_window_s: float = 300.0,  # 5-minute window
        max_decay_count: int = 5,       # After 5 repeats, minimum strength
        min_strength: float = 0.3,
        silent_probability_base: float = 0.3,
    ) -> None:
        self.decay_window_s = float(decay_window_s)
        self.max_decay_count = max(1, int(max_decay_count))
        self.min_strength = float(min_strength)
        self.silent_probability_base = float(silent_probability_base)
        # (scene_type, target_key) -> list of execution timestamps
        self._history: dict[tuple[str, str], list[float]] = defaultdict(list)

    def evaluate(
        self,
        scene_type: str,
        target_id: str | None,
    ) -> tuple[float, bool]:
        """Return ``(strength, use_voice)``.

        - ``strength``: 1.0 (first encounter) → ``min_strength`` (repeated).
        - ``use_voice``: True = full multimodal, False = visual-only.
        """
        now = monotonic()
        key = (scene_type, target_id or "__any__")
        self._prune(key, now)
        count = len(self._history[key])

        # Decay strength linearly.
        decay_ratio = min(count / self.max_decay_count, 1.0)
        strength = 1.0 - (1.0 - self.min_strength) * decay_ratio

        # Silent probability increases with repetition.
        silent_p = self.silent_probability_base + (decay_ratio * 0.4)
        use_voice = random.random() > silent_p

        return strength, use_voice

    def record(self, scene_type: str, target_id: str | None) -> None:
        now = monotonic()
        key = (scene_type, target_id or "__any__")
        self._history[key].append(now)

    def reset(self) -> None:
        self._history.clear()

    def _prune(self, key: tuple[str, str], now: float) -> None:
        cutoff = now - self.decay_window_s
        self._history[key] = [t for t in self._history[key] if t > cutoff]

File: Robot_Life_CPP/src/behavior/decay_tracker.cpp
#include "robot_life_cpp/behavior/decay_tracker.hpp"

#include <algorithm>
#include <chrono>

namespace robot_life_cpp::behavior {

BehaviorDecayTracker::BehaviorDecayTracker(
    double decay_window_s,
    int max_decay_count,
    double min_strength,
    double silent_probability_base)
    : decay_window_s_(std::max(1.0, decay_window_s)),
      max_decay_count_(std::max(1, max_decay_count)),
      min_strength_(std::clamp(min_strength, 0.0, 1.0)),
      silent_probability_base_(std::clamp(silent_probability_base, 0.0, 1.0)),
      rng_(std::random_device{}()) {}

std::pair<double, bool> BehaviorDecayTracker::evaluate(
    const std::string& scene_type, const std::string& target_id) {
  const auto key = key_for(scene_type, target_id);
  const auto now = now_mono();
  prune(key, now);
  const auto count = static_cast<double>(history_[key].size());
  const auto decay_ratio = std::min(count / static_cast<double>(max_decay_count_), 1.0);
  const auto strength = 1.0 - (1.0 - min_strength_) * decay_ratio;
  const auto silent_probability = std::clamp(silent_probability_base_ + decay_ratio * 0.4, 0.0, 1.0);
  std::uniform_real_distribution<double> dist(0.0, 1.0);
  const bool use_voice = dist(rng_) > silent_probability;
  return {strength, use_voice};
}

void BehaviorDecayTracker::record(const std::string& scene_type, const std::string& target_id) {
  const auto key = key_for(scene_type, target_id);
  const auto now = now_mono();
  prune(key, now);
  history_[key].push_back(now);
}

void BehaviorDecayTracker::reset() { history_.clear(); }

std::string BehaviorDecayTracker::key_for(const std::string& scene_type, const std::string& target_id) const {
  return scene_type + ":" + (target_id.empty() ? "__any__" : target_id);
}

void BehaviorDecayTracker::prune(const std::string& key, double now) {
  const auto cutoff = now - decay_window_s_;
  auto& series = history_[key];
  series.erase(std::remove_if(series.begin(), series.end(),
                              [cutoff](double ts) { return ts <= cutoff; }),
               series.end());
}

double BehaviorDecayTracker::now_mono() {
  using clock = std::chrono::steady_clock;
  return std::chrono::duration<double>(clock::now().time_since_epoch()).count();
}

}  // namespace robot_life_cpp::behavior
